torchtune/models/clip/_position_embeddings.py

```diff
@@ -0,0 +1,167 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+
+import torch
+from torch import nn
+
+# TODO (Felipe): add load hooks + interpolation on positional encodings,
+# so max_num_tiles can be variable and a trained model can be adapted to a
+# new value.
+
+
+class TokenPositionalEmbedding(nn.Module):
+    """
+    Token positional embedding for images. Different for every token (patch) in an image.
+
+    Args:
+        embed_dim (int): The dimensionality of each token embedding.
+        patch_grid_size (int): The side of a squared grid that represents how many
+             patches are in one tile, i.e. tile_size // patch_size.
+    """
+
+    def __init__(self, embed_dim: int, patch_grid_size: int) -> None:
+        super().__init__()
+        scale = embed_dim**-0.5
+        self.positional_embedding = nn.Parameter(
+            scale
+            * torch.randn((patch_grid_size**2 + 1, embed_dim))  # +1 for CLS token
+        )
+
+    def forward(self, x: torch.Tensor, *args) -> torch.Tensor:
+        """
+        args:
+            x (torch.Tensor): Tensor with shape (*, n_tokens, embed_dim)
+        returns:
+            torch.Tensor: The input tensor with added positional embeddings.
+        """
+        return x + self.positional_embedding
+
+
+class TiledTokenPositionalEmbedding(nn.Module):
+    """
+    Token positional embedding for tiled images. There are two positional embeddings in this module:
+        - local_token_positional_embedding: same for every tile, different for every token. Equivalent to
```

Comment:
Why is this called ```local_token_positional_embedding```? Isn't this exactly equivalent to the token positional embedding, both conceptually and in code?

Reply:
yes, the local is exactly the same as the token positional embedding, with the addition of the gate. Since this gate is shared, they have to be instantiated together. We cant untangle them.

eg:
x = x + local * gate
x = x + global * (1-gate)

Does it make sense? The only change I could make is to instantiate local_token_positional_embedding as TokenPositionalEmbedding, instead of nn.Parameter. Would you prefer that?

Reply:
they are not the same. If you have a 3 tiles, each with 2x2 patches:
[[1,2,3,4], [1,2,3,4], [1,2,3,4]]

the local_positional_embedding will be:
[[1,2,3,4], [1,2,3,4], [1,2,3,4]]

the global_token_positional_embedding will be:
[[1,2,3,4], [5,6,7,8], [9,10,11,12]]

Reply:
> they are not the same.

Sorry my question was about ```global_token_positional_embedding``` and ```TilePositionalEmbedding``` not ```global_token_positional_embedding``` and ```local_positional_embedding```. Are the first two different?

> calling .weight seems weird

Can't we just update the embedding module's forward to accept a scaling/gating tensor which defaults to ones? Sorry if that's not conceptually the same thing as the computation below and in that case disregard

Reply:
>> Sorry my question was about global_token_positional_embedding and TilePositionalEmbedding

If you have an image with 3 tiles, each with 2x2 patches:
[[1,2,3,4], [1,2,3,4], [1,2,3,4]]

the local_positional_embedding will be:
[[1,2,3,4], [1,2,3,4], [1,2,3,4]]

the global_token_positional_embedding will be:
[[1,2,3,4], [5,6,7,8], [9,10,11,12]]

TilePositionalEmbedding will be:
[[1,1,1,1], [2,2,2,2], [3,3,3,3]]

>> Can't we just update the embedding module's forward to accept a scaling/gating tensor which defaults to ones?

Unfortunately no, because this gating is also shared with the global_token_positional_embedding, which goes back to option 2 that you said you are not a fan of :P

"option 2) delete TokenPositionalEmbedding module and have a single module, with gating + global_token_positional_embedding as optional. This would add new flags/complexity, but less modules."

Reply:
Hmm, maybe I don't understand - let me ping you about this in a bit